FinnUI/Sources/Recycling/ProjectUnits/ProjectUnitsView.swift

```diff
@@ -0,0 +1,270 @@
+    }
+
+    private func configurePageControl(withNumberOfPages numberOfPages: Int) {
+        pageControl.numberOfPages = numberOfPages
+        pageControl.currentPage = 0
+    }
+}
+
+// MARK: - UICollectionViewDataSource
+
+extension ProjectUnitsView: UICollectionViewDataSource {
+    public func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        let numberOfItems = dataSource?.numberOfItems(inProjectUnitsView: self) ?? 0
+        configurePageControl(withNumberOfPages: numberOfItems)
+        return numberOfItems
+    }
+
+    public func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = collectionView.dequeue(ProjectUnitCell.self, for: indexPath)
+        guard let dataSource = dataSource else { return cell }
+
+        cell.remoteImageViewDataSource = remoteImageViewDataSource
+        cell.delegate = self
+
+        if let viewModel = dataSource.projectUnitsView(self, modelAtIndex: indexPath.row) {
+            cell.configure(with: viewModel)
+        }
+        cell.isFavorite = dataSource.projectUnitsView(self, unitAtIndexIsFavorite: indexPath.row)
+        return cell
+    }
+}
+
+// MARK: - UICollectionViewDelegate
+
+extension ProjectUnitsView: UICollectionViewDelegate {
+    public func scrollViewWillEndDragging(
+        _ scrollView: UIScrollView,
+        withVelocity velocity: CGPoint,
+        targetContentOffset: UnsafeMutablePointer<CGPoint>
+    ) {
+        guard isPagingEnabled else { return }
+
+        let targetOffsetX = targetContentOffset.pointee.x
+        let center = CGPoint(x: targetOffsetX + scrollView.frame.midX, y: scrollView.frame.midY)
+
+        if let indexPath = collectionView.indexPathForItem(at: center) {
+            pageControl.currentPage = indexPath.row
+        }
+    }
+
+    public func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        delegate?.projectUnitsView(self, didSelectUnitAtIndex: indexPath.row)
+    }
+}
+
+// MARK: - ProjectUnitCellDelegate
+
+extension ProjectUnitsView: ProjectUnitCellDelegate {
+    func projectUnitCell(_ projectUnitCell: ProjectUnitCell, didTapFavoriteButton button: UIButton) {
+        guard let indexPath = collectionView.indexPath(for: projectUnitCell) else { return }
+        delegate?.projectUnitsView(self, didTapFavoriteButton: button, forIndex: indexPath.row)
+    }
+}
+
+// MARK: - UICollectionViewFlowLayout
+
+private final class PagingCollectionViewLayout: UICollectionViewFlowLayout {
+    /// Returns the centered content offset to use after an animated layout update or change.
+    override func targetContentOffset(
+        forProposedContentOffset proposedContentOffset: CGPoint,
+        withScrollingVelocity velocity: CGPoint
+    ) -> CGPoint {
+        guard let bounds = collectionView?.bounds, let layoutAttributes = layoutAttributesForElements(in: bounds) else {
```

Comment:
i would put a comment here on what are you trying to achieve here.

why you need this? for what? i cannot tell much by just looking at the code

Reply:
Tbh, I'm not 100% sure. I copied `PagingCollectionViewLayout` from our `NeighborhoodProfile`, which also has paging, so I didn't have to implement the paging functionality myself.
But it looks like those variables are what we need to implement the paging, and if they're not present, we just pass it on to `UICollectionViewFlowLayout` so we'd just get a regular scroll. 🤔

Reply:
there is a `isPaginated` property in the collection view to achieve that behavior

Reply:
Thanks, will test and see if that works just as fine 😄

Reply:
I tried with `isPagingEnabled`, and it didn't behave as I wanted. `isPagingEnabled` is made for the cases where the cell width is the same as the collectionView's width, but for this view I want the previous and next cell to be visible as well. 😄

Reply:
I see, fair enough.

as a side note, this layout pagination I think can be achieved with the new uicollection view compositional layout, but ofc is iOS13+